keras/utils/data_utils.py

```diff
@@ -282,3 +295,322 @@ def validate_file(fpath, file_hash, algorithm='auto', chunk_size=65535):
+
+    def __init__(self, sequence, pickle_safe=False, scheduling='sequential'):
+        self.sequence = sequence
+        self.pickle_safe = pickle_safe
+        self.scheduling = scheduling
+        self.workers = 0
+        self.executor = None
+        self.queue = None
+        self.run_thread = None
+        self.stop_signal = None
+
+    def is_running(self):
+        return self.stop_signal is not None and not self.stop_signal.is_set()
+
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, workers could block on put())
+        """
+        if self.pickle_safe:
+            self.executor = multiprocessing.Pool(workers)
+        else:
+            self.executor = ThreadPool(workers)
+        self.queue = queue.Queue(max_q_size)
+        self.stop_signal = threading.Event()
+        self.run_thread = threading.Thread(target=self._run)
+        self.run_thread.daemon = True
+        self.run_thread.start()
+
+    def _run(self):
+        """ Function to submit request to the executor and queue the `Future` objects."""
+        sequence = list(range(len(self.sequence)))
+        while True:
+            if self.scheduling is not 'sequential':
+                random.shuffle(sequence)
+            for i in sequence:
+                if self.stop_signal.is_set():
+                    return
+                self.queue.put(
+                    self.executor.apply_async(get_index, (self.sequence, i)), block=True)
+
+    def get(self):
+        """Create a generator to extract data from the queue. Skip the data if it's None.
+
+        # Returns
+            Generator yielding tuples (inputs, targets)
+                or (inputs, targets, sample_weights)
+        """
+        try:
+            while self.is_running():
+                inputs = self.queue.get(block=True).get()
+                if inputs is not None:
+                    yield inputs
+        except Exception as e:
+            self.stop()
+            raise StopIteration(e)
+
+    def stop(self, timeout=None):
+        """Stop running threads and wait for them to exit, if necessary.
+
+        Should be called by the same thread which called start().
+
+        # Arguments
+            timeout: maximum time to wait on thread.join()
+        """
+        self.stop_signal.set()
+        with self.queue.mutex:
+            self.queue.queue.clear()
+            self.queue.unfinished_tasks = 0
+            self.queue.not_full.notify()
+        self.executor.close()
+        self.run_thread.join(timeout)
+
+
+class GeneratorEnqueuer(SequenceEnqueuer):
+    """Builds a queue out of a data generator.
+
+    Used in `fit_generator`, `evaluate_generator`, `predict_generator`.
+
+    # Arguments
+        generator: a generator function which endlessly yields data
+        pickle_safe: use multiprocessing if True, otherwise threading
```

Comment:
`pickle_safe` is generally not a good API argument, it is not very descriptive and is Python-specific (the Keras API should generally avoid being Python-specific). This is an opportunity to get rid of it.

I would use `use_multiprocessing` (or `multiprocessing`, but then we have to handle to name collision with the module name) in the new code, and deprecate the `pickle_safe` argument in the generator methods at a later date.